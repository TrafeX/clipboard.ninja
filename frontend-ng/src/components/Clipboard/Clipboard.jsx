import {
  Card,
  CardBody,
  CardHeader,
  Heading,
  SimpleGrid,
  Text,
  useColorModeValue,
} from "@chakra-ui/react";
import { useContext } from "react";
import { SocketContext } from "../../context/SocketContext";
import Connect from "../Connect";
import Receive from "../Receive";
import Send from "../Send";
// import { useMatomo } from '@datapunt/matomo-tracker-react';

const Clipboard = ({
  ownRoomNumber,
  connectedToRoom,
  usersInRoom,
  messages,
}) => {
  const socket = useContext(SocketContext);

  // const { trackEvent } = useMatomo();

  const sendMessage = (message) => {
    socket.emit("publish", message);
    const event = {
      category: "user-interaction",
      action: "publish-message",
    };
    // trackEvent(event);
  };

  const connectToRoom = (room) => {
    socket.emit("join", room);
    const event = {
      category: "user-interaction",
      action: "connect-to-device",
    };
    // trackEvent(event);
  };

  const cardBgColor = useColorModeValue("white", "gray.700");

  return (
    <SimpleGrid columns={{ base: 1, md: 2 }} spacing={8} padding={8}>
      <Card
        bg={cardBgColor}
        boxShadow={"lg"}
        p={{ sm: 3, md: 5 }}
        rounded={"md"}
      >
        <CardHeader pb={{ base: 0, lg: 1 }}>
          <Heading size="md">Connect & send</Heading>
          <Text fontSize={"sm"}>
            Open this app on both devices and connect to the other device by
            entering the Device ID
          </Text>
        </CardHeader>
        <CardBody>
          <Connect
            ownRoomNumber={ownRoomNumber}
            connectedToRoom={connectedToRoom}
            connectToRoom={connectToRoom}
          />
          <Send
            connectedToRoom={connectedToRoom}
            sendMessage={sendMessage}
            usersInRoom={usersInRoom}
          />
        </CardBody>
      </Card>
      <Card boxShadow={"lg"} p={5} rounded={"md"}>
        <CardHeader pb={{ base: 0, lg: 1 }}>
          <Heading size="md">Received text</Heading>
          <Text fontSize={"sm"}>
            Once connected, the text you receive from the other device will be
            shown here
          </Text>
        </CardHeader>
        <CardBody>
          <Receive receivedMessages={messages} />
        </CardBody>
      </Card>
    </SimpleGrid>
  );
};

export default Clipboard;
