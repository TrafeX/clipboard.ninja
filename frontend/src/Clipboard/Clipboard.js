// @flow
import React, { Component } from 'react';
import Send from './../Send';
import Receive from './../Receive';
import Connect from './../Connect';
import Status from './../Status';
import {blueGrey400} from 'material-ui/styles/colors';
import {Card, CardTitle, CardText} from 'material-ui/Card';
import {io} from 'socket.io-client';
import ReactGA from 'react-ga';

const style = {
  cardLeft: {
    flex: '1 1 400px',
    textAlign: 'center',
    margin: 10,
  },
  cardRight: {
    flex: '1 1 400px',
    textAlign: 'center',
    margin: 10,
  },
  footer: {
    color: 'white',
    margin: 10,
    paddingLeft: 10,
  },
  a: {
    color: 'white',
  },
  background: {
    backgroundColor: blueGrey400,
    paddingBottom: 10,
    height: 240,
  }
};
type Props = {};
type State = {
  messages: Array<string>,
  ownRoomNumber: ?number,
  connectedToRoom: ?number,
  status: string,
  usersInRoom: number,
};

class Clipboard extends Component<Props, State> {
  socket: Object;

  state = {
    messages: [],
    ownRoomNumber: null,
    connectedToRoom: null,
    status: '',
    usersInRoom: 0,
  };

  componentDidMount = () => {

    // @todo: Move this to higer component to keep the socket alive?
    this.socket = io(process.env.REACT_APP_BACKEND_URL);

    this.socket.on('connect_error', () => {
      this.setState({
        status: 'Connection to server failed',
        ownRoomNumber: null,
      })
    });
    this.socket.on('message', (message: string) => {
      const msgs = this.state.messages;
      msgs.unshift(message);

      this.setState({
        messages: msgs,
      });
    });
    this.socket.on('registered', (room: number) => {
      this.setState({
        ownRoomNumber: room,
      });
    });
    this.socket.on('subscribed', (room: number, usersInRoom: number) => {
      this.setState({
        status: `Connected to device with ID ${room}`,
        ownRoomNumber: room,
        connectedToRoom: room,
        usersInRoom: usersInRoom,
      });
    });
    this.socket.on('unsubscribed', (usersInRoom: number) => {
      if (usersInRoom <= 1) {
        // Only one left in room
        this.setState({
          status: 'The other device is gone',
          connectedToRoom: null,
          usersInRoom: usersInRoom,
        });
      }
      this.setState({
        usersInRoom: usersInRoom,
      });
    });
    this.socket.on('deviceid-not-exists', () => {
      this.setState({
        status: 'Device ID doesn\'t exists. Enter the device ID of the other device',
        connectedToRoom: null,
        // @todo: Trigger error on Connect input element
      });
    });
  };

  componentWillUnmount = () => {
    this.socket.disconnect();
  };

  sendMessage = (message: string) => {
    this.socket.emit('publish', message);
    ReactGA.event({
      category: 'user-interaction',
      action: 'publish-message'
    });
  };

  connectToRoom = (room: ?number) => {
    this.socket.emit('join', room);
    ReactGA.event({
      category: 'user-interaction',
      action: 'connect-to-device'
    });
  };

  render() {
    return (
      <div className="container">
        <Status
          status={this.state.status}
        />
        <Card style={style.cardLeft}>
          <CardTitle
            title="Connect & Send"
            subtitle="First you need to connect to the other device by entering the Device ID"
            />
          <CardText style={{paddingTop: 0}}>
            <Connect
              ownRoomNumber={this.state.ownRoomNumber}
              connectedToRoom={this.state.connectedToRoom}
              connectToRoom={this.connectToRoom}
              />
            <Send
              connectedToRoom={this.state.connectedToRoom}
              sendMessage={this.sendMessage}
              usersInRoom={this.state.usersInRoom}
              />
          </CardText>
        </Card>
        <Card style={style.cardRight}>
          <CardTitle
            title="Received text"
            subtitle="The text you send to and receive from the other device will be shown here"
            />
          <CardText>
            <Receive
              receivedMessages={this.state.messages}
              />
          </CardText>
        </Card>
      </div>
    );
  }
}

export default Clipboard;
