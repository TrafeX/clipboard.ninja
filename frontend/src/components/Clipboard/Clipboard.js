import React, { useContext } from 'react';
import Send from '../Send';
import Receive from '../Receive';
import Connect from '../Connect';
import { blueGrey400 } from 'material-ui/styles/colors';
import { Card, CardTitle, CardText } from 'material-ui/Card';
import ReactGA from 'react-ga';
import { SocketContext } from '../../context/SocketContext';

const style = {
  cardLeft: {
    flex: '1 1 400px',
    textAlign: 'center',
    margin: 10,
  },
  cardRight: {
    flex: '1 1 400px',
    textAlign: 'center',
    margin: 10,
  },
  footer: {
    color: 'white',
    margin: 10,
    paddingLeft: 10,
  },
  a: {
    color: 'white',
  },
  background: {
    backgroundColor: blueGrey400,
    paddingBottom: 10,
    height: 240,
  }
};

const Clipboard = ({ownRoomNumber, connectedToRoom, usersInRoom, messages}) => {

  const socket = useContext(SocketContext);

  // componentWillUnmount = () => {
  //   this.socket.disconnect();
  // };

  const sendMessage = (message: string) => {
    socket.emit('publish', message);
    ReactGA.event({
      category: 'user-interaction',
      action: 'publish-message'
    });
  };

  const connectToRoom = (room: ?number) => {
    socket.emit('join', room);
    ReactGA.event({
      category: 'user-interaction',
      action: 'connect-to-device'
    });
  };

  return (
    <div className="container">
      <Card style={style.cardLeft}>
        <CardTitle
          title="Connect & Send"
          subtitle="First you need to connect to the other device by entering the Device ID"
        />
        <CardText style={{ paddingTop: 0 }}>
          <Connect
            ownRoomNumber={ownRoomNumber}
            connectedToRoom={connectedToRoom}
            connectToRoom={connectToRoom}
          />
          <Send
            connectedToRoom={connectedToRoom}
            sendMessage={sendMessage}
            usersInRoom={usersInRoom}
          />
        </CardText>
      </Card>
      <Card style={style.cardRight}>
        <CardTitle
          title="Received text"
          subtitle="The text you send to and receive from the other device will be shown here"
        />
        <CardText>
          <Receive
            receivedMessages={messages}
          />
        </CardText>
      </Card>
    </div>
  );
}

export default Clipboard;
