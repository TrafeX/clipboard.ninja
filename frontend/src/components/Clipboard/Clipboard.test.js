import React from 'react';
import Clipboard from './Clipboard';
import { shallow } from 'enzyme';

it('renders without crashing', () => {
  const wrapper = shallow(<Clipboard />);
  expect(wrapper).toMatchSnapshot();
});
