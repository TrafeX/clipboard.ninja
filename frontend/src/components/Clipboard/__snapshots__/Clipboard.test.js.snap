// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="container"
>
  <Card
    expandable={false}
    expanded={null}
    initiallyExpanded={false}
    style={
      Object {
        "flex": "1 1 400px",
        "margin": 10,
        "textAlign": "center",
      }
    }
  >
    <CardTitle
      subtitle="First you need to connect to the other device by entering the Device ID"
      title="Connect & Send"
    />
    <CardText
      style={
        Object {
          "paddingTop": 0,
        }
      }
    >
      <MuiComponent
        connectToRoom={[Function]}
      />
      <MuiComponent
        sendMessage={[Function]}
      />
    </CardText>
  </Card>
  <Card
    expandable={false}
    expanded={null}
    initiallyExpanded={false}
    style={
      Object {
        "flex": "1 1 400px",
        "margin": 10,
        "textAlign": "center",
      }
    }
  >
    <CardTitle
      subtitle="The text you send to and receive from the other device will be shown here"
      title="Received text"
    />
    <CardText>
      <Receive />
    </CardText>
  </Card>
</div>
`;
